### ISSUE TITLE
Secure Custom Domain Support (TLS) - Linksharing

### ISSUE THREAD

**Issue by ferristocrat** [storj/roadmap#39]
### What is the problem/pain point?
The Storj Linksharing Service does not support HTTPS for custom domains. Current documentation for using a custom domain with Linkshare: https://docs.storj.io/dcs/how-tos/host-a-static-website/host-a-static-website-with-the-cli-and-linksharing-service/  
### What is the Impact?
Customers who want to provide access to objects in Storj buckets to their own end users from a web frontend use Linksharing. When they want (or need) to have a custom domain that is secured with TLS for the content served, the url must be proxied through a CDN such as Cloudflare or BunnyCDN. A couple problems unfold as a result:

1. Proxying certain types of content, such as streaming video, is against the terms of service for providers such as Cloudflare – so these customers are stuck
2. CDN providers (without a TOS restriction) cache content, reducing the amount of egress from Storj and adding an unnecessary third party

Storj DCS customers often have strict requirements from their own customers, including strict security reviews and policies. There are customers who have to whitelabel every domain that can be accessed, and the security teams won’t approve a generic hosting domain, such as linkshare, because understandably unblocking that opens up more than they want.

### Why now?
This particular pain point is trending among existing and new customers.

### User Story
> **_As a Storj DCS user, I want to host content stored in DCS on my own domain via https, so my content is delivered securely and with a link my customers will be comfortable with._**

### Acceptance Criteria

- User can define a custom domain for linksharing
    - Storj generates certificate on behalf of users
    - User has a way to prove ownership of domain to generate the certificate
- Linkshare URLs generated for this particular customer should have an additional URL, that contains their custom domain

### Out of Scope
  - "Bring your own certificate"
    - Customers might want to bring their own certificate rather than rely on the autogenerated LetsEncrypt cert
    - In this case, users would upload their own certificate (such as a DigiCert certificate)
    - This will be saved for future consideration on the roadmap

## Measures of Success
**Egress**

- Increased Egress via Linksharing

**New Customers**

- Increased # customers using Linksharing with Custom Domains

## Project Milestones
Project Initiation: https://github.com/storj/gateway-mt/milestone/8
Research: https://github.com/storj/gateway-mt/milestone/11
Limits: https://github.com/storj/gateway-mt/milestone/12
TLS Termination: https://github.com/storj/gateway-mt/milestone/13
Backend Work: https://github.com/storj/gateway-mt/milestone/14
Observability: https://github.com/storj/gateway-mt/milestone/15
Paid Tier: https://github.com/storj/storj/milestone/31

## Project Blueprint
[Blueprint: Linksharing Secure Custom Domain Support (TLS)](https://github.com/storj/gateway-mt/blob/main/docs/blueprints/tls-custom-domains.md)

**Comment by shaupt131** [comment]
## Week 31 update

**In Review:**

- Blueprint has preliminary approval
- Blueprint is out for final approval

**In Progress:**

- Breaking down blueprint into milestones and issues

**Comment by amwolff** [comment]
## Week 35 update

**Completed:**

- Blueprint is approved and has been broken down into milestones and issues

**In Progress:**

- Research
    - Reviewing CertMagic library
- Observability
    - Establishing metrics that are necessary for ensuring success
    - Establishing metrics for number of customers currently using custom domains

**Comment by shaupt131** [comment]
## Week 39 update

**Completed:**
- Research
    - Review of CertMagic library
- Observability
    - Established metrics that are necessary for ensuring success
    - Establishing metrics for number of customers currently using custom domains
- Limits
    - Comparing different certificate providers

**In Progress:**
- Research
    - Research how existing customers will be impacted/migrated to TLS custom domains
- Backend
    - Setting up storage backend for CertMagic
- Limits
    - Requesting higher limits from certificate providers

**Comment by shaupt131** [comment]
## Week 41 update

**Completed:**
- Research
    - Research how existing customers will be impacted/migrated to TLS custom domains
- Observability
    - Determine metrics needed

**In Progress:**
- TLS Termination
    - Wire up CertMagic in linksharing
- Backend
    - Setting up storage backend for CertMagic
    - Implement distributed locks for the storage backend for CertMagic
- Limits
    - Requesting higher limits from certificate providers

**Comment by storjrobot** [comment]
This issue has been mentioned on **Storj Community Forum (official)**. There might be relevant details there:

https://forum.storj.io/t/static-website-hosting-malformed-request/20075/4

**Comment by davidwolfpaw** [comment]
After the Google Safe Browsing issues, this has risen in importance for our usage of Storj for our Mastodon instance.
